subsys/bluetooth/host/id.c

```diff
@@ -345,15 +384,48 @@ int bt_id_set_adv_private_addr(struct bt_le_ext_adv *adv)
 }
 #endif /* defined(CONFIG_BT_PRIVACY) */
 
-static void adv_update_rpa(struct bt_le_ext_adv *adv, void *data)
+#if defined(CONFIG_BT_EXT_ADV) && defined(CONFIG_BT_PRIVACY)
+static void adv_disable_rpa(struct bt_le_ext_adv *adv, void *data)
 {
+	uint8_t adv_index = bt_le_ext_adv_get_index(adv);
+	bool *adv_disabled = data;
+	bool rpa_invalid = true;
+
+	adv_disabled[adv_index] = false;
+
+	/* Invalidate RPA for non-limited advertising sets. */
+	if (atomic_test_bit(adv->flags, BT_ADV_LIMITED)) {
+		return;
+	}
+
+	atomic_clear_bit(adv->flags, BT_ADV_RPA_VALID);
+
+	/* Disable advertising sets to prepare them for RPA update. */
 	if (atomic_test_bit(adv->flags, BT_ADV_ENABLED) &&
-	    !atomic_test_bit(adv->flags, BT_ADV_LIMITED) &&
 	    !atomic_test_bit(adv->flags, BT_ADV_USE_IDENTITY)) {
-		int err;
-
 		bt_le_adv_set_enable_ext(adv, false, NULL);
 
+		adv_disabled[adv_index] = true;
+	}
+
+	/* Notify the user about the RPA timeout and set the RPA validity. */
+	if (adv->cb && adv->cb->rpa_timed_out) {
+		rpa_invalid = adv->cb->rpa_timed_out(adv);
+	}
+
+	if (!rpa_invalid) {
+		atomic_set_bit(adv->flags, BT_ADV_RPA_VALID);
+	}
```

Comment:
Similar situation here: There's not really any reason to clear the `BT_ADV_RPA_VALID` flag, then ask if it should be cleared, and if not, then re-set it, instead of just asking and only if requested then clear the flag.

Reply:
Sure, I can simplify this